modules/gateway/nodes.go

```diff
@@ -32,45 +31,26 @@ func (g *Gateway) addNode(addr modules.NetAddress) error {
 	return nil
 }
 
-// managedAddUntrustedNode adds an address to the set of nodes on the network, but
-// first verifies that there is a reachable node at the provided address.
-func (g *Gateway) managedAddUntrustedNode(addr modules.NetAddress) error {
-	// Performing the ping during testing does not work.
-	if build.Release == "testing" {
-		g.mu.Lock()
-		defer g.mu.Unlock()
-		return g.addNode(addr)
-	}
-
-	// Ping the untrusted node to see whether or not there's acutally a
+// pingNode verifies that there is a reachable node at the provided address
+// by performing the Sia gateway handshake protocol.
+func (g *Gateway) pingNode(addr modules.NetAddress) error {
+	// Ping the untrusted node to see whether or not there's actually a
 	// reachable node at the provided address.
 	conn, err := g.dial(addr)
 	if err != nil {
 		return err
 	}
+	defer conn.Close()
 	// If connection succeeds, supply an unacceptable version so that we
 	// will not be added as a peer.
 	//
 	// NOTE: this is a somewhat clunky way of specifying that you didn't
 	// actually want a connection.
-	encoding.WriteObject(conn, "0.0.0")
-	var reject string
-	err = encoding.ReadObject(conn, &reject, build.MaxEncodedVersionLength)
-	if err != nil {
-		g.log.Debugln("ERROR: version handshake ping terminated unexpectedly:", err)
-	}
-	if reject != "reject" {
-		g.log.Debugln("WARN: peer does not seem to have correctly rejected our ping:", reject)
-	}
-	conn.Close()
-
-	g.mu.Lock()
-	defer g.mu.Unlock()
-	err = g.addNode(addr)
-	if err != nil {
-		return err
+	_, err = connectVersionHandshake(conn, "0.0.0")
+	if err != errPeerRejectedConn {
+		g.log.Debugln("ERROR: version handshake should have resulted in rejection; got", err)
 	}
-	return g.save()
+	return nil
```

Comment:
oops, thought this comment was on `managedAcceptConnNewPeer`. You are right, I think we want to check for the specific error.
However, it's probably alright if the handshake returns `nil`. That just means the remote peer is being a little weird and not rejecting peers that advertise as 0.0.0. There might be a legitimate reason to do so.